Implement IndexOrDocValuesQuery for geo_shape field

iverase opened the pull request:
IndexOrDocValuesQuery is an optimisation for queries when there are in a filter with another query that is more restrictive. In that case, it might be faster to just check the matches for that query directly with the doc values instead of traversing the BKD index.

With the introduction of doc values for geo_shape, it is possible now to implement  such query for this type of data.

iverase committed:
Implement IndexOrDocValuesQuery for geo_shape field

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Geo)

iverase committed:
Merge branch 'master' into IndexOrDocValues

jpountz commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java:
```diff
@@ -0,0 +1,169 @@
+            indexQuery = LatLonShape.newGeometryQuery(FIELD_NAME, ShapeField.QueryRelation.DISJOINT, q);
+            docValQuery = new LatLonShapeDocValuesQuery(FIELD_NAME, ShapeField.QueryRelation.DISJOINT, q);
+            assertEquals(s.count(indexQuery), s.count(docValQuery));
+            // CONTAINS and multi-shapes fails due to LUCENE-9595
```
do you think it should block this PR?

iverase replied:
I think we can wait until we depend on a Lucene version that contains the fix.

jpountz commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java:
```diff
@@ -0,0 +1,169 @@
+                doc.add(docVal);
+                w.addDocument(doc);
+            } catch (Exception e) {
+                // ignore
```
is it safe? Can we narrow down the set of exceptions types we ignore?

iverase replied:
I remove the try / catch

jpountz commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:
```diff
@@ -0,0 +1,125 @@
+        if (field == null) {
+            throw new IllegalArgumentException("field must not be null");
+        }
+        this.field = field;
```
should you also test that `geometry` has a length > 0?

jpountz replied:
or alternatively rewrite the query when length == 0?

iverase replied:
This is done in `VectorGeoShapeWithDocValuesQueryProcessor`. It gets rewritten to a `MatchNoDocsQuery`, is that enough?

jpountz reviewed (commented):
I didn't review the Component2DVisitor logic, but the way the query gets built makes sense to me.

talevy reviewed (commented):
Do you think it would be helpful to have both index and doc-value queries execute against the same tests instead of just  having doc-value queries be considered correct if they behave like index-queries? Since theoretically the counts may be the same, but maybe they are including different documents? highly unlikely, but just a thought.

iverase committed:
Merge branch 'master' into IndexOrDocValues

iverase committed:
Address review comments and fix some bugs on the way

iverase commented:
Thanks @Tal. I updated the test so we actually check the hits are the same. I improved the test as well which surface and issue with Contains queries and geometry collections.

I added more documentation to Component2DVisitor, let me know if that's ok.

iverase committed:
Merge branch 'master' into IndexOrDocValues

iverase committed:
clean up test

iverase committed:
unused imports

iverase commented:
We upgraded Elasticsearch to a Lucene 8.8 snapshot so this PR is not blocked any more. Could you have another look @talevy ?

talevy commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:
```diff
@@ -0,0 +1,193 @@
+
+                    @Override
+                    public float matchCost() {
+                        return 1000f; // TODO: what should it be?
```
I am not too familiar with this. So long as `matchCost` is constant and the ConstantScoreScorer is used, this number is arbitrary and has no effect, right?

iverase replied:
My understanding is that this value is used when you have more that one `TwoPhaseIterator` in a conjunction in order to sort them. You want to execute the cheaper iterators first.

In put case we do not have a good estimation of cost so it is good enough to return a big number. Note that this is done the same for LatLonPointDocValues.

talevy replied:
thanks!

talevy reviewed (approved):
LGTM pending resolution

iverase committed:
Merge branch 'master' into IndexOrDocValues

iverase committed:
compile error

iverase committed:
checkStyle
